azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/message_queue.py

```diff
@@ -0,0 +1,129 @@
+from six.moves import queue
+import logging
+from threading import Event
+from azure.iot.hub.devicesdk.transport import constant
+
+logger = logging.getLogger(__name__)
+
+
+class MessageQueue(
+    queue.Queue, object
+):  # add object as multiple inheritance b/c Queue is old style (no inheritance from object)
+    def __init__(self, on_enable, on_disable):
+        super(MessageQueue, self).__init__()
+        self._enabled = False
+        self._on_enable = on_enable
+        self._on_disable = on_disable
+
+    def enable(self):
+        if not self._enabled:
+            self._on_enable()  # must do this before changing status!
+            self._enabled = True
+
+    def disable(self):
+        if self._enabled:
+            self._enabled = False
+            self._on_disable()  # must do this after changing status!
+
+    @property
+    def enabled(self):
+        return self._enabled
+
+    @enabled.setter
+    def enabled(self, v):
+        if v:
+            self.enable()
+        else:
+            self.disable()
```

Comment:
I'm seeing 2 ways of doing the same thing (call enable/disable methods or change the property). I would rather only have one way. if we can't make this read-only, then I would make enable/disable private (underscore methods)

Reply:
enable() and disable() are the preferred ways to do this, so should not be private. This setter is more of an insurance policy - the "enabled" property is not able to be private, as it's important to check the enabled status of a MessageQueue. However, because it's public, it could potentially be overwritten, which would result in unexpected and hard to trace behaviour when the enabled property is set without triggering the on_enable/on_disable handlers.

We could just throw an exception on an attempt to set here instead, but I felt that this was a smoother solution

---
In reply to: [259534214](https://github.com/Azure/azure-iot-sdk-python-preview/pull/49#discussion_r259534214) [](ancestors = 259534214)